Include memcheck test playbook into CI suite.

dargasudarshan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you -->

**What this PR does / why we need it**:
This PR adds the test suite to check the memory consumed by JIVA controller pod with dd workloads.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #1342

**Special notes for your reviewer**:
This playbook will run couple of dd workload threads and verify the memory consumed by the OpenEBS JIVA controller pod.

As part of verifying memory usage, getting used memory for interval of 20s for 10 iterations.
A watermark of 500MB has been set based on benchmark runs with the workload profile chosen in this test.
#### TODO: Identify better memory consumption strategies.

ksatchit commented on e2e/ansible/playbooks/hyperconverged/test-k8s-memleak/test-mem.py:
```diff
@@ -0,0 +1,31 @@
+else:
+        print "Fail"
+
+
```
@dargasudarshan , couple of observations:

- While free is an indicator of memory consumption, let us try using pmap which seems to be an indicator of the *type of memory* used. This maybe good checks to have in the testcase. However, it might also be a good idea to parameterize the type of check itself - Having said that, if this is checking the mem issue - it should be ok to implement/track the suggestion as a separate issue.

- While using python/bash scripts might be inevitable in cases and ansible does allow it, it might be a good idea to check if the same can be accomplished via ansible itself. For ex:  Get memory via : 

```
kubectl exec pvc-c71e49dc-273a-11e8-b04f-000c298ff5fc-ctrl-5f659cb984-5x8qx -c pvc-c71e49dc-273a-11e8-b04f-000c298ff5fc-ctrl-con -- pmap -x 1 | awk ''/total'/ {print $3}'
```
...along with a do/until/retry/failed_when block for the check.

ksatchit commented:
@dargasudarshan , can you please paste the output of the ansible playbook run with verbosity level -vv

ksatchit commented on e2e/ansible/playbooks/hyperconverged/test-k8s-memleak/memleak.yml:
```diff
@@ -0,0 +1,163 @@
+       - name: Place data using dd
+         shell: timeout 200s dd if=/dev/urandom of={{mount_point}}/f1 bs=512k count=10000 &
+         become: true
+         delegate_to: "{{groups['kubernetes-kubeminions'].0}}"
```
@dargasudarshan , is it possible to parameterize the workload characteristics, (blocksize, count etc.,) While this is OK, we can probably alter-it/pass it as a runtime var if needed w/ defaults as the ones already used.

dargasudarshan replied:
Yes @ksatchit I do changes accordingly

gprasath commented:
@ksatchit when we run this playbook against jiva:0.5.1, we can catch the failure as shown below:

```
TASK [Run python script for memory usage validation it takes couple of minutes] ****************************************************************************************
task path: /home/giri/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-memleak/memleak.yml:143
Using module file /usr/lib/python2.7/dist-packages/ansible/modules/commands/command.py
<172.28.128.4> ESTABLISH SSH CONNECTION FOR USER: ubuntu
<172.28.128.4> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o ControlPath=/home/giri/.ansible/cp/ad01abce11 172.28.128.4 '/bin/sh -c '"'"'echo ~ && sleep 0'"'"''
<172.28.128.4> (0, '/home/ubuntu\n', '')
<172.28.128.4> ESTABLISH SSH CONNECTION FOR USER: ubuntu
<172.28.128.4> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o ControlPath=/home/giri/.ansible/cp/ad01abce11 172.28.128.4 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo /home/ubuntu/.ansible/tmp/ansible-tmp-1521013451.1-39029817292676 `" && echo ansible-tmp-1521013451.1-39029817292676="` echo /home/ubuntu/.ansible/tmp/ansible-tmp-1521013451.1-39029817292676 `" ) && sleep 0'"'"''
<172.28.128.4> (0, 'ansible-tmp-1521013451.1-39029817292676=/home/ubuntu/.ansible/tmp/ansible-tmp-1521013451.1-39029817292676\n', '')
<172.28.128.4> PUT /tmp/tmpQ2vYIV TO /home/ubuntu/.ansible/tmp/ansible-tmp-1521013451.1-39029817292676/command.py
<172.28.128.4> SSH: EXEC sftp -b - -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=10 -o ControlPath=/home/giri/.ansible/cp/ad01abce11 '[172.28.128.4]'
<172.28.128.4> (0, 'sftp> put /tmp/tmpQ2vYIV /home/ubuntu/.ansible/tmp/ansible-tmp-1521013451.1-39029817292676/command.py\n', '')
<172.28.128.4> ESTABLISH SSH CONNECTION FOR USER: ubuntu
<172.28.128.4> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o ControlPath=/home/giri/.ansible/cp/ad01abce11 172.28.128.4 '/bin/sh -c '"'"'chmod u+x /home/ubuntu/.ansible/tmp/ansible-tmp-1521013451.1-39029817292676/ /home/ubuntu/.ansible/tmp/ansible-tmp-1521013451.1-39029817292676/command.py && sleep 0'"'"''
<172.28.128.4> (0, '', '')
<172.28.128.4> ESTABLISH SSH CONNECTION FOR USER: ubuntu
<172.28.128.4> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=10 -o StrictHostKeyChecking=no -o ControlPath=/home/giri/.ansible/cp/ad01abce11 -tt 172.28.128.4 '/bin/sh -c '"'"'sudo -H -S  -p "[sudo via ansible, key=jgdpuofjlpooeglwuuojixtfeoseejzg] password: " -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-jgdpuofjlpooeglwuuojixtfeoseejzg; /usr/bin/python /home/ubuntu/.ansible/tmp/ansible-tmp-1521013451.1-39029817292676/command.py; rm -rf "/home/ubuntu/.ansible/tmp/ansible-tmp-1521013451.1-39029817292676/" > /dev/null 2>&1'"'"'"'"'"'"'"'"' && sleep 0'"'"''
<172.28.128.4> (0, '\r\n\r\n{"changed": true, "end": "2018-03-14 07:48:15.546265", "stdout": "745.83984375 MB\\n786.484375 MB\\n787.38671875 MB\\n786.41796875 MB\\n787.35546875 MB\\n786.32421875 MB\\n343.11328125 MB\\n350.74609375 MB\\n350.8046875 MB\\n350.34765625 MB\\n[745.83984375, 786.484375, 787.38671875, 786.41796875, 787.35546875, 786.32421875, 343.11328125, 350.74609375, 350.8046875, 350.34765625]\\nFail", "cmd": "python test-mem.py", "rc": 0, "start": "2018-03-14 07:44:15.070424", "stderr": "", "delta": "0:04:00.475841", "invocation": {"module_args": {"warn": true, "executable": null, "_uses_shell": true, "_raw_params": "python test-mem.py", "removes": null, "creates": null, "chdir": null, "stdin": null}}}\r\n', 'Shared connection to 172.28.128.4 closed.\r\n')
fatal: [localhost -> None]: FAILED! => {
    "changed": true,
    "cmd": "python test-mem.py",
    "delta": "0:04:00.475841",
    "end": "2018-03-14 07:48:15.546265",
    "failed_when_result": true,
    "invocation": {
        "module_args": {
            "_raw_params": "python test-mem.py",
            "_uses_shell": true,
            "chdir": null,
            "creates": null,
            "executable": null,
            "removes": null,
            "stdin": null,
            "warn": true
        }
    },
    "rc": 0,
    "start": "2018-03-14 07:44:15.070424",
    "stderr": "",
    "stderr_lines": [],
    "stdout": "745.83984375 MB\n786.484375 MB\n787.38671875 MB\n786.41796875 MB\n787.35546875 MB\n786.32421875 MB\n343.11328125 MB\n350.74609375 MB\n350.8046875 MB\n350.34765625 MB\n[745.83984375, 786.484375, 787.38671875, 786.41796875, 787.35546875, 786.32421875, 343.11328125, 350.74609375, 350.8046875, 350.34765625]\nFail",
    "stdout_lines": [
        "745.83984375 MB",
        "786.484375 MB",
        "787.38671875 MB",
        "786.41796875 MB",
        "787.35546875 MB",
        "786.32421875 MB",
        "343.11328125 MB",
        "350.74609375 MB",
        "350.8046875 MB",
        "350.34765625 MB",
        "[745.83984375, 786.484375, 787.38671875, 786.41796875, 787.35546875, 786.32421875, 343.11328125, 350.74609375, 350.8046875, 350.34765625]",
        "Fail"
    ]
}
```

dargasudarshan commented:
The results with the openeba/jiva:latest are here:

TASK [Place data using dd] *********************************************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-memleak/memleak.yml:120
changed: [localhost -> None] => {"changed": true, "cmd": "timeout 200s dd if=/dev/urandom of=/mnt/jiva/f1 bs=128k count=10000 &", "delta": "0:00:01.003327", "end": "2018-03-14 11:50:26.411449", "rc": 0, "start": "2018-03-14 11:50:25.408122", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

TASK [Place data using dd] *********************************************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-memleak/memleak.yml:125
changed: [localhost -> None] => {"changed": true, "cmd": "timeout 200s dd if=/dev/urandom of=/mnt/jiva/f2 bs=512k count=10000 &", "delta": "0:00:01.008927", "end": "2018-03-14 11:50:27.649810", "rc": 0, "start": "2018-03-14 11:50:26.640883", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

TASK [Get $HOME of Kube minion for kubernetes user] ********************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-memleak/memleak.yml:130
changed: [localhost -> None] => {"changed": true, "cmd": "source ~/.profile; echo $HOME", "delta": "0:00:00.024665", "end": "2018-03-14 11:50:28.139906", "rc": 0, "start": "2018-03-14 11:50:28.115241", "stderr": "", "stderr_lines": [], "stdout": "/home/ubuntu", "stdout_lines": ["/home/ubuntu"]}

TASK [Copy the python script to kubeminion] ****************************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-memleak/memleak.yml:137
ok: [localhost -> None] => {"changed": false, "checksum": "52f925ebdda9b7c502dfb3884c9b8ee30310b2fc", "gid": 1000, "group": "ubuntu", "mode": "0664", "owner": "ubuntu", "path": "/home/ubuntu/test-mem.py", "size": 638, "state": "file", "uid": 1000}

TASK [Run python script for memory usage validation it takes couple of minutes] ****************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-memleak/memleak.yml:143
changed: [localhost -> None] => {"changed": true, "cmd": "python test-mem.py", "delta": "0:03:20.619373", "end": "2018-03-14 11:53:50.540270", "failed_when_result": false, "rc": 0, "start": "2018-03-14 11:50:29.920897", "stderr": "", "stderr_lines": [], "stdout": "[313.1171875, 313.19140625, 311.640625, 312.62890625, 312.73828125, 312.12890625, 312.01953125, 314.40234375, 313.0390625, 312.8515625]\nPass", "stdout_lines": ["[313.1171875, 313.19140625, 311.640625, 312.62890625, 312.73828125, 312.12890625, 312.01953125, 314.40234375, 313.0390625, 312.8515625]", "Pass"]}

TASK [set_fact] ********************************************************************************************************************************************************
task path: /home/devops/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-memleak/memleak.yml:150
ok: [localhost] => {"ansible_facts": {"flag": "PASS"}, "changed": false}

gila commented on e2e/ansible/playbooks/hyperconverged/test-k8s-memleak/test-mem.py:
```diff
@@ -0,0 +1,31 @@
+    count = count + 1
+    #Obtaining memory consumed using free command from the node where the controller
+    #is scheduled.
+    cmd_used_mem = "sudo free | awk ' /'Mem'/ {print $3}'"
```
You are grabbing system memory not process memory.

dargasudarshan replied:
@gila I will do this task as enhancement as part of next sprint and track under https://github.com/openebs/openebs/issues/1350

dargasudarshan replied:
@gila Now modified the code to grab process memory.

payes commented:
Can we also try Valgrind.

dargasudarshan commented:
@payes For valgrind you have raised this issue https://github.com/mayadata-io/cstor-tracking/issues/98. We can track through this.

dargasudarshan commented:
The results with the openeba/jiva:latest after implementing the review comments:


TASK [Copy the python script to kubeminion] ****************************************************************************************************************************
task path: /home/devops/build/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-memleak/memleak.yml:128
ok: [localhost -> None] => {"changed": false, "checksum": "96e84c83387c83fe842c13f8b6266e702d82d0c2", "gid": 1000, "group": "ubuntu", "mode": "0664", "owner": "ubuntu", "path": "/home/ubuntu/test-mem.py", "size": 1407, "state": "file", "uid": 1000}

TASK [Run python script for memory usage validation it takes couple of minutes] ****************************************************************************************
task path: /home/devops/build/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-memleak/memleak.yml:134
changed: [localhost -> None] => {"changed": true, "cmd": "python test-mem.py", "delta": "0:01:42.363381", "end": "2018-03-16 11:06:38.341548", "failed_when_result": false, "rc": 0, "start": "2018-03-16 11:04:55.978167", "stderr": "", "stderr_lines": [], "stdout": "490.4921875 MB\n490.4921875 MB\n490.4921875 MB\n490.4921875 MB\n490.4921875 MB\n[490.4921875, 490.4921875, 490.4921875, 490.4921875, 490.4921875]\nPass", "stdout_lines": ["490.4921875 MB", "490.4921875 MB", "490.4921875 MB", "490.4921875 MB", "490.4921875 MB", "[490.4921875, 490.4921875, 490.4921875, 490.4921875, 490.4921875]", "Pass"]}

TASK [set_fact] ********************************************************************************************************************************************************
task path: /home/devops/build/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-memleak/memleak.yml:140
ok: [localhost] => {"ansible_facts": {"flag": "PASS"}, "changed": false}

TASK [Logout to iSCSI target] ******************************************************************************************************************************************
task path: /home/devops/build/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-memleak/memleak-cleanup.yml:2
changed: [localhost -> None] => {"changed": true, "connection_changed": true, "nodes": ["iqn.2016-09.com.openebs.jiva:pvc-c17981ab-2909-11e8-87f7-02b983f0a4db", "iqn.2016-09.com.openebs.jiva:pvc-896b1aa2-2909-11e8-87f7-02b983f0a4db", "iqn.2016-09.com.openebs.jiva:pvc-863ab822-2908-11e8-87f7-02b983f0a4db"]}

TASK [Delete the iSCSI target] *****************************************************************************************************************************************
task path: /home/devops/build/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-memleak/memleak-cleanup.yml:11
changed: [localhost -> None] => {"changed": true, "cmd": "source ~/.profile; kubectl delete -f test-pvc.yaml", "delta": "0:00:00.155784", "end": "2018-03-16 11:06:41.343017", "rc": 0, "start": "2018-03-16 11:06:41.187233", "stderr": "", "stderr_lines": [], "stdout": "persistentvolumeclaim \"vut\" deleted", "stdout_lines": ["persistentvolumeclaim \"vut\" deleted"]}

TASK [Unmount the ext4 filesystem] *************************************************************************************************************************************
task path: /home/devops/build/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-memleak/memleak-cleanup.yml:17
changed: [localhost -> None] => {"changed": true, "dump": "0", "fstab": "/etc/fstab", "name": "/mnt/jiva", "opts": "defaults", "passno": "0"}

TASK [Confirm the iSCSI target has been deleted] ***********************************************************************************************************************
task path: /home/devops/build/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-memleak/memleak-cleanup.yml:24
changed: [localhost -> None] => {"attempts": 1, "changed": true, "cmd": "source ~/.profile; kubectl get pvc", "delta": "0:00:00.156680", "end": "2018-03-16 11:06:42.133655", "rc": 0, "start": "2018-03-16 11:06:41.976975", "stderr": "No resources found.", "stderr_lines": ["No resources found."], "stdout": "", "stdout_lines": []}

TASK [Send slack notification] *****************************************************************************************************************************************
task path: /home/devops/build/openebs/e2e/ansible/playbooks/hyperconverged/test-k8s-memleak/memleak.yml:151
skipping: [localhost] => {"changed": false, "skip_reason": "Conditional result was False"}
META: ran handlers
META: ran handlers

PLAY RECAP *************************************************************************************************************************************************************
localhost                  : ok=25   changed=17   unreachable=0    failed=0

ksatchit commented:
With the process memory being tracked now, there are few more cases we might want to handle in enhancements to this test: 
- Garbage Collected, but memory still w/ runtime (how to handle this case)
- Run dd w/ sync flags
- Reduce the logic in the python script/use ansible.

These will be tracked under a separate issue. Merging this to get it running on CI asap.
